File: android/app/src/main/kotlin/org/gaindrive/android/data/cache/PinRepository.kt
package org.gaindrive.android.data.cache

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.gaindrive.android.data.ServerRegistry
import org.gaindrive.android.data.SettingsStore
import org.gaindrive.android.data.StreamUrls
import org.gaindrive.android.data.local.LocalLibrary
import org.gaindrive.android.data.local.PinDao
import org.gaindrive.android.data.local.PinEntity
import org.gaindrive.android.data.model.AudioFormat
import org.gaindrive.android.data.model.AudioQuality
import org.gaindrive.android.data.model.ItemRef
import org.gaindrive.android.data.model.Song
import org.gaindrive.android.playback.playableAudioFor
import javax.inject.Inject
import javax.inject.Singleton

/** Why a pin was refused, or that it was not. */
sealed interface PinResult {
	data object Ok : PinResult

	/** Pinning this would put more than the cap out of eviction's reach. */
	data class TooLarge(val neededBytes: Long, val capBytes: Long) : PinResult

	/** Nothing stored to download yet - open it once while online first. */
	data object NotKnownYet : PinResult
}

/**
 * What the user has asked to keep, and everything that follows from it:
 * protection from eviction, and downloading the audio now rather than
 * whenever it next happens to be played.
 */
@Singleton
class PinRepository @Inject constructor(
	private val pinDao: PinDao,
	private val local: LocalLibrary,
	private val pinnedKeys: PinnedKeys,
	private val downloads: DownloadQueue,
	private val audioCache: AudioCache,
	private val settings: SettingsStore,
	private val registry: ServerRegistry,
	private val streamUrls: StreamUrls,
	private val pinnedArt: PinnedArt,
	private val artDownloader: ArtDownloader,
	private val scope: CoroutineScope,
) {

	private val _pins = MutableStateFlow<List<Pin>>(emptyList())

	/** The pins themselves, as placed - an album pin stays one entry. */
	val pins: StateFlow<List<Pin>> = _pins.asStateFlow()

	private val _coverage = MutableStateFlow(PinCoverage(emptyMap()))

	private val _protectedKeys = MutableStateFlow<Set<String>>(emptySet())

	/** The songs those pins expand to, for the "downloaded" markers. */
	val protectedKeys: StateFlow<Set<String>> = _protectedKeys.asStateFlow()

	/**
	 * How far each pin has got, keyed by its encoded ref.
	 *
	 * Progress is counted in whole tracks, not bytes: Media3 pushes download
	 * state changes but not continuous progress, and the cache's own notion of
	 * "stored" is per track anyway. For an album that reads naturally - three of
	 * twelve - and for a single track it is the difference between a spinner and
	 * a tick, which is all that was missing.
	 */
	val statuses: StateFlow<Map<String, PinStatus>> = combine(
		_coverage,
		audioCache.cachedKeys,
		downloads.states,
	) { coverage, stored, downloadStates ->
		// A finished download counts even when the cache cannot vouch for it -
		// see DownloadStates.completed on why it often cannot.
		val here = stored + downloadStates.completed
		coverage.byPin.mapValues { (_, keys) ->
			PinStatus(
				stored = keys.count { it in here },
				total = keys.size,
				phase = pinPhaseOf(keys, here, downloadStates),
			)
		}
	}.stateIn(scope, SharingStarted.Eagerly, emptyMap())

	/**
	 * Keeps an existing download library from being re-fetched on upgrade.
	 *
	 * Before quality was a setting, everything downloaded was the original file.
	 * Letting the new default apply to an install that already has pins would
	 * quietly re-download all of them at Opus, over whatever connection happens
	 * to be there. Recording what those bytes actually are instead leaves the
	 * choice to the user, who can change it in Settings and get the dialog.
	 *
	 * Runs before the pin collector below, so protection is never applied under
	 * the wrong quality even briefly.
	 */
	private suspend fun adoptExistingDownloadQuality() {
		if (settings.audioQualityChosen.first()) return
		if (pinDao.all().isEmpty()) return
		settings.setAudioQuality(AudioQuality.ORIGINAL)
	}

	@OptIn(FlowPreview::class)
	private fun start(scope: CoroutineScope) {
		scope.launch {
			adoptExistingDownloadQuality()
			pinDao.observeAll().collect { rows ->
				// A pin for a server that has since been removed protects a key
				// nothing will ever play. Dropped here rather than from the
				// removal itself: the registry cannot call into this without
				// closing a loop, and removing the rows re-runs this collector
				// with them gone.
				val known = registry.servers.first().map { it.id }.toSet()
				val (live, orphans) = rows.mapNotNull { it.toPin() }
					.partition { it.ref.server in known }
				orphans.map { it.ref.server }.distinct()
					.forEach { pinDao.removeForServer("${it.value}/%") }
				applyProtection(live)
			}
		}
		// A pinned playlist gains a track, or a pinned album's track list is
		// seen for the first time: what the pin covers has changed even though
		// the pin has not. Debounced because a browse writes several tables in
		// a burst and only the settled result matters.
		scope.launch {
			local.revision.drop(1).debounce(REVISION_DEBOUNCE_MS).collect {
				applyProtection(_pins.value)
			}
		}
	}

	init {
		start(scope)
	}

	fun isPinned(ref: ItemRef): Boolean = _pins.value.any { it.ref == ref }

	/**
	 * Pins [ref] and starts downloading what it covers.
	 *
	 * Refused when it would not fit: eviction cannot reclaim pinned bytes, so a
	 * pin past the cap silently turns the cap into a lie. Better to say so here,
	 * where raising the cap is one screen away.
	 */
	suspend fun pin(ref: ItemRef, kind: PinKind): PinResult {
		val songs = songsFor(Pin(ref, kind))
		if (songs.isEmpty()) return PinResult.NotKnownYet

		val cap = settings.cacheMaxBytes.first()
		val quality = settings.audioQuality.first()
		val alreadyPinned = audioCache.pinnedBytes.value
		val adding = songs
			.filterNot {
				pinnedKeys.isPinned(CacheKeys.of(it.ref, it.effectiveQuality(quality)))
			}
			.sumOf { it.storedSizeAt(quality) }
		if (alreadyPinned + adding > cap) {
			return PinResult.TooLarge(alreadyPinned + adding, cap)
		}

		pinDao.add(PinEntity(refKey = ref.encode(), kind = kind.name))
		// Protection is applied here rather than waiting for the DAO's flow: a
		// download starting before its key is protected could be evicted by the
		// very bytes it is writing.
		applyProtection(_pins.value + Pin(ref, kind))
		enqueue(songs)
		return PinResult.Ok
	}

	suspend fun unpin(ref: ItemRef) {
		val pin = _pins.value.firstOrNull { it.ref == ref } ?: return
		pinDao.remove(ref.encode())

		val before = _protectedKeys.value
		applyProtection(_pins.value - pin)
		// Only what nothing else still covers: a track pinned on its own and
		// also part of a pinned album must survive the album being unpinned.
		(before - _protectedKeys.value).forEach { downloads.remove(it) }
	}

	/**
	 * Re-enqueues what a pin covers, for a download that failed.
	 *
	 * Already-stored tracks cost nothing to re-request: the download manager
	 * finds them complete in the cache and reports them done. Removal is not
	 * offered here - the pin list in Settings is where a download you have
	 * given up on gets deleted.
	 */
	suspend fun retry(ref: ItemRef) {
		val pin = _pins.value.firstOrNull { it.ref == ref } ?: return
		enqueue(songsFor(pin))
	}

	suspend fun refresh() = applyProtection(_pins.value)

	/**
	 * Pins with something readable attached, for the list in Settings.
	 *
	 * A pin whose item is no longer in the mirror still appears, named by its
	 * kind alone - it is the only place it can be removed from, so hiding it
	 * would strand it.
	 */
	suspend fun describe(pins: List<Pin>): List<PinnedItem> = pins.map { pin ->
		val label = when (pin.kind) {
			PinKind.SONG -> local.song(pin.ref)?.title
			PinKind.ALBUM -> local.album(pin.ref)?.let { "${it.artistName} - ${it.title}" }
			PinKind.PLAYLIST -> local.playlists(pin.ref.server)
				.firstOrNull { it.ref == pin.ref }?.name
		}
		PinnedItem(pin, label ?: "Unknown ${pin.kind.name.lowercase()}")
	}

	private suspend fun applyProtection(pins: List<Pin>) {
		val expansion = expand(pins)
		val coverage = expansion.coverage
		val keys = coverage.allKeys
		// Two forms of the same set, and they are no longer interchangeable.
		//
		// The evictor works in *cache* keys, which carry the quality, because
		// that is what the stored bytes are filed under. Everything above -
		// availability marks, pin coverage, the UI - works in bare refs, which
		// is what the library mirror and the DAO understand.
		//
		// Only the quality currently set is protected: a copy left behind by an
		// earlier setting stays playable but becomes ordinary evictable cache,
		// which is what makes a quality change reclaim its own space.
		val quality = settings.audioQuality.first()
		// A video's soundtrack is filed under the substituted quality (see
		// AudioQuality.forVideoAudio), so protecting it under the plain one
		// would name bytes that do not exist and leave the ones that do
		// evictable - a downloaded film that quietly stops playing offline.
		// One bulk query rather than a lookup per ref: a playlist pin can
		// cover hundreds.
		val videos = local.videoRefKeys(keys)
		// Set before publishing: the evictor reads this, and a download racing
		// ahead of it would be evictable for the length of the race.
		pinnedKeys.keys = keys.mapTo(mutableSetOf()) {
			CacheKeys.of(it, if (it in videos) quality.forVideoAudio() else quality)
		}
		_pins.value = pins
		_coverage.value = coverage
		_protectedKeys.value = keys
		// After the pins, not before: nothing evicts art, so there is no race
		// to lose here, and this reaches the disk and the network. Holding the
		// pin state behind it would make every pin feel slower than it is.
		applyArt(expansion)
	}

	/**
	 * The picture half of a pin: which files to keep, and which to go and get.
	 *
	 * Skipped wholesale when the expansion retained anything, because a
	 * retained pass is one where the mirror was emptied underneath it (see
	 * [retaining]). The art refs it produced would be a handful of fallbacks
	 * rather than the real set, and acting on them would delete the covers of
	 * everything the user pinned. Leaving the index and the files exactly as
	 * they are costs one stale pass and the next one puts it right.
	 */
	private suspend fun applyArt(expansion: Expansion) {
		if (expansion.retained) return
		pinnedArt.publish(artDownloader.index(expansion.artRefs))
		pinnedArt.retain(expansion.artRefs.mapTo(mutableSetOf()) { ArtKeys.fileName(it) })
		// Launched rather than awaited: pinning must not wait on pictures, and
		// this is also the pass that picks up a portrait the server had not
		// resolved when the pin was placed. Republished afterwards because the
		// index holds only files that exist, and these did not a moment ago.
		scope.launch {
			artDownloader.fetchMissing(expansion.artRefs)
			pinnedArt.publish(artDownloader.index(expansion.artRefs))
		}
	}

	/** Re-fetches art a pin covers but does not have, for "Clear cover art". */
	suspend fun refreshArt() = applyProtection(_pins.value)

	/**
	 * Re-fetches everything pinned, at whatever quality is now set.
	 *
	 * Protection is re-applied first so `PinnedKeys` holds the new keys before
	 * any download starts writing under them - the same ordering [pin] relies
	 * on, for the same reason.
	 */
	suspend fun refreshDownloads() {
		applyProtection(_pins.value)
		_pins.value.forEach { enqueue(songsFor(it)) }
	}

	/**
	 * What a set of pins covers, in audio and in pictures.
	 *
	 * Both at once because they need the same rows, and reading them twice
	 * would be two passes over the mirror for every pin change. [retained]
	 * says whether [retaining] had to carry anything over, which is the signal
	 * that the mirror was empty underneath this pass and that acting on the
	 * art would be destructive.
	 */
	private class Expansion(
		val coverage: PinCoverage,
		val artRefs: List<ItemRef>,
		val retained: Boolean,
	)

	private suspend fun expand(pins: List<Pin>): Expansion {
		// Read once for the whole pass rather than per pin: a flip halfway
		// through would otherwise produce a coverage that included the videos
		// under some pins and not others.
		val audioOnly = settings.videoAudioOnly.first()
		val songs = pins.associateWith { songsOf(it) }
		val fresh = expandPins(
			pins = pins,
			albumSongs = songs.filterKeys { it.kind == PinKind.ALBUM }
				.map { (pin, rows) -> pin.ref to rows.downloadable(audioOnly).map(Song::ref) }
				.toMap(),
			playlistSongs = songs.filterKeys { it.kind == PinKind.PLAYLIST }
				.map { (pin, rows) -> pin.ref to rows.downloadable(audioOnly).map(Song::ref) }
				.toMap(),
		)
		// The same condition [retaining] acts on, asked before it does: a pin
		// that covered something and now covers nothing means the mirror went
		// out from under this pass.
		val previous = _coverage.value.byPin
		val retained = fresh.byPin.any { (pin, covered) ->
			covered.isEmpty() && !previous[pin].isNullOrEmpty()
		}
		val coverage = retaining(fresh)
		// Every song, not only the downloadable ones: a film excluded from the
		// audio a pin covers is still an album the user will look at offline.
		val art = pins.flatMap { pin ->
			artRefsOf(
				pin = pin,
				album = if (pin.kind == PinKind.ALBUM) local.album(pin.ref) else null,
				songs = songs[pin].orEmpty(),
			)
		}.distinct()
		return Expansion(coverage, art, retained)
	}

	private suspend fun songsOf(pin: Pin): List<Song> = when (pin.kind) {
		PinKind.SONG -> listOfNotNull(local.song(pin.ref))
		PinKind.ALBUM -> local.songsOfAlbum(pin.ref)
		PinKind.PLAYLIST -> local.songsOfPlaylist(pin.ref)
	}

	/**
	 * Keeps the previous membership of any pin that has just come back empty.
	 *
	 * An empty expansion is a legitimate *state* - pinning an album that has
	 * never been opened protects nothing, which `expandPins` documents - but it
	 * is never a legitimate *transition* for a pin that already covered
	 * something. What produces one is the mirror being emptied underneath it:
	 * removing a server, or changing one's browse mode. Both write through
	 * `LocalLibrary`, which bumps the revision this class listens on, so the
	 * recompute lands while the tables are bare.
	 *
	 * Without this the coverage collapses, `PinnedKeys` shrinks, and the evictor
	 * is free to reclaim audio the user explicitly asked to keep - silently, and
	 * only until they next happen to open the album. The stale membership is the
	 * safer of the two wrong answers: it protects bytes that are already on
	 * disk, and the next successful browse replaces it.
	 */
	private fun retaining(fresh: PinCoverage): PinCoverage {
		val previous = _coverage.value.byPin
		if (previous.isEmpty()) return fresh
		return PinCoverage(
			fresh.byPin.mapValues { (pin, songs) ->
				songs.ifEmpty { previous[pin].orEmpty() }
			}
		)
	}

	private suspend fun songsFor(pin: Pin): List<Song> =
		songsOf(pin).downloadable(settings.videoAudioOnly.first())

	private suspend fun enqueue(songs: List<Song>) {
		songs.forEach { song ->
			// A video only ever reaches here through downloadable(), which lets
			// one past exactly when its soundtrack is what will be fetched - so
			// `isVideo` is the flag the URL builder wants, not a second reading
			// of the setting that could disagree with the first.
			// The same declaration playback makes, and it must be: both build
			// the same cache key, so a pinned copy and a played one that
			// declared differently would write different bytes under it.
			val target = streamUrls.forDownload(
				song.ref,
				song.isVideo,
				::playableAudioFor,
			) ?: return@forEach
			downloads.add(song.ref, target)
		}
	}

	/**
	 * The quality this song's bytes are actually filed under.
	 *
	 * A video's soundtrack is always a transcode - `AudioFormat.ORIGINAL` for
	 * one means "send the film", which is not a thing that can be stored - so
	 * the substitution has to be applied wherever a cache key or a size is
	 * derived, not only where the URL is built.
	 */
	private fun Song.effectiveQuality(quality: AudioQuality): AudioQuality =
		if (isVideo) quality.forVideoAudio() else quality

	/**
	 * Roughly what this song will occupy once stored.
	 *
	 * [Song.sizeBytes] is the size of the file on the server, which is the wrong
	 * number as soon as anything is transcoded - an album of FLACs would be
	 * refused for a pin that actually fits several times over at Opus. Bitrate
	 * times duration is close enough for a capacity check.
	 *
	 * A video is the extreme case of the same thing: its `sizeBytes` is the
	 * whole film, several gigabytes for something that stores as a hundred
	 * megabytes, which would refuse a pin that fits many times over.
	 */
	private fun Song.storedSizeAt(quality: AudioQuality): Long {
		val effective = effectiveQuality(quality)
		return if (effective.format == AudioFormat.ORIGINAL) sizeBytes
		else (duration * effective.bitRate * 125L)
	}

	private fun PinEntity.toPin(): Pin? {
		val ref = ItemRef.decode(refKey) ?: return null
		val pinKind = runCatching { PinKind.valueOf(kind) }.getOrNull() ?: return null
		return Pin(ref, pinKind)
	}
}

File: ios/GainDrive/Data/Cache/PinRepository.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// What the user asked to keep, what that expands to, and how much of it is
/// here.
///
/// The pins themselves are persisted as a **Codable file, not a database**:
/// this is tens of rows of intent, and the thing that genuinely needs one is
/// the metadata mirror, which is a later stage. Adding a dependency for a list
/// this small would be a dependency taken for the wrong reason.
@MainActor
@Observable
final class PinRepository {
	private(set) var pins: [Pin] = []
	private(set) var usageBytes: Int64 = 0
	/// A refusal worth showing - today only "this would not fit". Cleared when
	/// the screen that showed it says so.
	private(set) var message: String?

	private var membership: [Pin.ID: [ItemRef]] = [:]
	/// Pinned **and** here. Everything else on disk is `cached`, which the
	/// store answers for directly - there is no list of it, because the list
	/// would be out of date the moment eviction ran.
	private var stored: Set<ItemRef> = []
	/// Everything the store holds, pinned or not - a track kept by playing it
	/// belongs to no pin, so it is in nothing else this class computes.
	///
	/// **Read from the store, never asked for per row.** A row cannot await the
	/// disk while it is being drawn, and asking during a view update would
	/// mutate observable state mid-render.
	private var held: Set<ItemRef> = []
	private var progress: [ItemRef: Double] = [:]
	private var failed: Set<ItemRef> = []

	@ObservationIgnored private let library: LibraryRepository
	@ObservationIgnored private let store: AudioStore
	@ObservationIgnored private let queue: DownloadQueue
	@ObservationIgnored private let targets: StreamTargets
	@ObservationIgnored private let settings: SettingsStore
	@ObservationIgnored private let file: URL

	init(
		library: LibraryRepository, store: AudioStore, queue: DownloadQueue,
		targets: StreamTargets, settings: SettingsStore
	) {
		self.library = library
		self.store = store
		self.queue = queue
		self.targets = targets
		self.settings = settings
		let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
		self.file = base.appending(path: "pins.json")
		load()

		queue.onProgress = { [weak self] ref, fraction in
			self?.progress[ref] = fraction
			self?.failed.remove(ref)
		}
		queue.onFinished = { [weak self] ref in
			guard let self else { return }
			self.progress[ref] = nil
			self.failed.remove(ref)
			self.stored.insert(ref)
			Task { await self.reloadHeld() }
		}
		queue.onFailed = { [weak self] ref in
			self?.progress[ref] = nil
			self?.failed.insert(ref)
		}
	}

	// MARK: - Reading

	func isPinned(_ ref: ItemRef, kind: PinKind) -> Bool {
		pins.contains { $0.ref == ref && $0.kind == kind }
	}

	/// One track's state.
	///
	/// **`stored` and `cached` are not the same claim.** A pinned track is safe
	/// from eviction; one kept because it was played can go tonight, and
	/// collapsing them would promise a permanence the second does not have.
	func state(for song: ItemRef) -> DownloadState {
		if stored.contains(song) { return .stored }
		if let fraction = progress[song] { return .running(fraction: fraction) }
		if failed.contains(song) { return .failed }
		if held.contains(song) { return .cached }
		return .absent
	}

	/// The store telling us what it holds has changed - a track finished
	/// arriving, or eviction took one. Wired in the composition root.
	func storeChanged() {
		Task { await reloadHeld() }
	}

	private func reloadHeld() async {
		held = await store.heldRefs()
		usageBytes = await store.totalBytes()
	}

	/// A whole pin's state, which is what the album screen's control shows.
	///
	/// **It reports what is happening, not what was asked for** - the reason
	/// `android/SCREENS.md` gives for the same control. A pin that is half here
	/// is running, not done, and one whose last track failed is failed however
	/// many succeeded.
	func state(of pin: Pin) -> DownloadState {
		let songs = Pins.expand([pin], membership: membership)
		guard !songs.isEmpty else {
			// Pinned, but never resolved - a pin whose album has not been read
			// yet protects nothing, and saying "here" would be a lie.
			return .running(fraction: 0)
		}
		if songs.contains(where: { failed.contains($0) }) { return .failed }
		let here = songs.filter { stored.contains($0) }.count
		if here == songs.count { return .stored }
		return .running(fraction: Double(here) / Double(songs.count))
	}

	func clearMessage() {
		message = nil
	}

	// MARK: - Writing

	func toggle(_ pin: Pin) async {
		if isPinned(pin.ref, kind: pin.kind) {
			await remove(pin)
		} else {
			await add(pin)
		}
	}

	private func add(_ pin: Pin) async {
		// A song pin is its own membership and needs no read, which is also
		// why it is the one kind that cannot fail to resolve.
		let songs = pin.kind == .song ? [] : await resolve(pin)
		guard pin.kind == .song || !songs.isEmpty else {
			message = "Could not read what that holds. Try again when the server answers."
			return
		}
		let refs = pin.kind == .song ? [pin.ref] : songs.map(\.ref)

		// **Refused rather than allowed to overrun.** Nothing here is evicted,
		// so pinning past the cap would quietly turn the cap into a lie -
		// `android/CACHING.md` reaches the same rule from the other direction,
		// where eviction exists but cannot reclaim pinned bytes.
		//
		// A song pin contributes nothing to the estimate, since its size was
		// never read. That errs towards letting one track through rather than
		// refusing it for want of a number, which is the right direction for a
		// figure that is an estimate either way.
		let needed = songs.filter { !stored.contains($0.ref) }.reduce(Int64(0)) {
			$0 + Pins.estimatedBytes(of: $1, quality: settings.audioQuality)
		}
		// **Against the pinned bytes, not the used bytes.** Everything else on
		// disk was kept because it was played and gives way to eviction, so
		// counting it would refuse a pin that fits perfectly well.
		let room = settings.cacheCapBytes - (await store.protectedBytes())
		guard needed <= room else {
			message = """
				Not enough room. That needs about \(Self.readable(needed)) \
				and \(Self.readable(max(room, 0))) is left. Raise the limit in \
				Storage, or remove something.
				"""
			return
		}

		pins.append(pin)
		if pin.kind != .song { membership[pin.id] = refs }
		persist()
		await pushLimits()
		await startMissing(refs)
	}

	func remove(_ pin: Pin) async {
		let before = Pins.expand(pins, membership: membership)
		pins.removeAll { $0.id == pin.id }
		membership = Pins.pruned(membership, to: pins)
		persist()

		// **Unpinning removes the bytes**, which is what "remove download" is
		// taken to mean and what stops orphaned files nothing lists. Only the
		// songs no *other* pin still covers: an album and a playlist may hold
		// the same track.
		let after = Pins.expand(pins, membership: membership)
		await pushLimits()
		for ref in before.subtracting(after) {
			queue.cancel(ref)
			progress[ref] = nil
			failed.remove(ref)
			stored.remove(ref)
			await store.remove(ref)
		}
		await reloadHeld()
	}

	/// Everything, including the pins - there is nothing on disk that is not
	/// pinned, so a flush that kept them would re-download immediately.
	func removeEverything() async {
		queue.cancelAll()
		pins = []
		membership = [:]
		progress = [:]
		failed = []
		stored = []
		persist()
		await pushLimits()
		await store.removeAll()
		await reloadHeld()
	}

	// MARK: - Refresh

	/// The cap changed in Settings; the store enforces it, so it has to hear.
	func capChanged() async {
		await pushLimits()
	}

	/// Re-reads what each pin covers and fetches anything missing.
	///
	/// This is what makes a pinned playlist cover a track added since - the
	/// reason a pin records intent rather than a song list. Run at launch and
	/// whenever the Storage screen appears.
	func refresh() async {
		await syncStored()
		progress = await queue.inFlight()
		queue.resume()

		var resolved: [Pin.ID: [ItemRef]] = [:]
		for pin in pins where pin.kind != .song {
			resolved[pin.id] = await resolve(pin).map(\.ref)
		}
		membership = Pins.pruned(
			Pins.merging(membership, resolved: resolved), to: pins)
		persist()
		await startMissing(Array(Pins.expand(pins, membership: membership)))
	}

	// MARK: - Machinery

	/// Empty for a song pin, which is its own membership - `Pins.expand` puts
	/// the ref straight in, so there is nothing to read and nothing to store.
	private func resolve(_ pin: Pin) async -> [Song] {
		switch pin.kind {
		case .song:
			return []
		// **Videos are never downloaded**, whatever they are pinned inside. One
		// film evicts the whole stored library, and a re-encoded one arrives
		// with no `Content-Length` so completeness could never be established.
		// Dropped here rather than at the download, so a mostly-audio album
		// pins cleanly and its film is simply not part of what the pin covers.
		case .album:
			return ((try? await library.albumDetail(pin.ref))?.songs ?? [])
				.filter { !$0.isVideo }
		case .playlist:
			return ((try? await library.playlist(pin.ref))?.songs ?? [])
				.filter { !$0.isVideo }
		}
	}

	private func startMissing(_ refs: [ItemRef]) async {
		for ref in refs where !stored.contains(ref) && progress[ref] == nil {
			// The same declaration playback makes, and it must be the same:
			// `StreamTargets` is shared precisely so a download and a play
			// cannot ask for different bytes, and they derive one cache key.
			guard
				let target = await targets.target(
					for: ref, playable: avfoundationPlayable(for:))
			else { continue }
			progress[ref] = 0
			queue.start(ref, quality: target.quality, url: target.url)
		}
	}

	private func syncStored() async {
		let wanted = Pins.expand(pins, membership: membership)
		stored = await store.storedRefs(among: wanted)
		await pushLimits()
		await reloadHeld()
	}

	/// Tells the store what it may not evict, and how much room it has.
	///
	/// Pushed rather than pulled: the store knows nothing about pins and should
	/// not learn, and this class already computes the set for its own answers.
	private func pushLimits() async {
		await store.setLimits(
			cap: settings.cacheCapBytes,
			protecting: Pins.expand(pins, membership: membership))
	}

	// MARK: - Persistence

	private struct Record: Codable {
		var pins: [Pin] = []
		var membership: [String: [ItemRef]] = [:]
	}

	private func load() {
		guard let data = try? Data(contentsOf: file),
			let record = try? JSONDecoder().decode(Record.self, from: data)
		else { return }
		pins = record.pins
		membership = record.membership
	}

	private func persist() {
		let record = Record(pins: pins, membership: membership)
		guard let data = try? JSONEncoder().encode(record) else { return }
		try? data.write(to: file, options: .atomic)
	}

	static func readable(_ bytes: Int64) -> String {
		ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
	}
}

extension Pins {
	/// What a track is likely to take at this quality.
	///
	/// **Not `sizeBytes`**, which is the size of the file as stored. Asking for
	/// AAC 160 of a FLAC album would then overestimate by five or six times and
	/// refuse a pin that fits comfortably. The rate is the one actually being
	/// requested, so the arithmetic is the same one the server will do.
	static func estimatedBytes(of song: Song, quality: AudioQuality) -> Int64 {
		guard quality.format != .original else { return Int64(song.sizeBytes) }
		guard song.duration > 0 else { return Int64(song.sizeBytes) }
		return Int64(song.duration) * Int64(quality.bitRate) * 1000 / 8
	}
}
